src/libAtomVM/nifs.c

```diff
@@ -3165,6 +3172,133 @@ static term nif_binary_split(Context *ctx, int argc, term argv[])
     return result_list;
 }
 
+static bool get_binary_scope_slice(term binary, term options, term scope_atom, BinaryPosLen *scope_slice)
+{
+    term scope_opt = interop_proplist_get_value_default(options, scope_atom, term_invalid_term());
+    if (term_is_invalid_term(scope_opt)) {
+        size_t size = term_binary_size(binary);
+        return term_normalize_binary_pos_len(binary, 0, (avm_int_t) size, scope_slice);
+    }
+
+    if (UNLIKELY(!term_is_tuple(scope_opt) || term_get_tuple_arity(scope_opt) != 2)) {
+        return false;
+    }
+
+    term pos_term = term_get_tuple_element(scope_opt, 0);
+    term len_term = term_get_tuple_element(scope_opt, 1);
+    if (UNLIKELY(!term_is_any_integer(pos_term) || !term_is_any_integer(len_term))) {
+        return false;
+    }
+
+    avm_int_t pos = term_to_int(pos_term);
+    avm_int_t len = term_to_int(len_term);
+    return term_normalize_binary_pos_len(binary, pos, len, scope_slice);
+}
+
+static bool is_valid_pattern(term t)
+{
+    if (term_is_binary(t)) {
+        return term_binary_size(t) > 0;
+    }
+
+    if (!term_is_nonempty_list(t)) {
+        return false;
+    }
+
+    while (term_is_nonempty_list(t)) {
+        term pattern_term = term_get_list_head(t);
+        if (UNLIKELY(!term_is_binary(pattern_term))) {
+            return false;
+        }
+        if (UNLIKELY(term_binary_size(pattern_term) == 0)) {
+            return false;
+        }
+        t = term_get_list_tail(t);
+    }
+    bool proper = term_is_nil(t);
+    if (UNLIKELY(!proper)) {
+        return false;
+    }
+    return true;
+}
+
+static BinaryPosLen find_pattern_in_binary(term binary_term, BinaryPosLen scope_slice, term pattern_term)
+{
+    const char *binary = term_binary_data(binary_term) + scope_slice.pos;
+    size_t size = scope_slice.len;
+    const char *pattern = term_binary_data(pattern_term);
+    size_t pattern_size = term_binary_size(pattern_term);
+
+    BinaryPosLen pattern_slice = term_invalid_binary_pos_len();
+    const char *sub_binary = memmem(binary, size, pattern, pattern_size);
```

Comment:
do we need to access the content of sub_binary using `[ ]`? if not we can just use `const void *`.

Reply:
While I kinda agree, `void *` for me means that I'm dealing with something generic and I really need to be aware of pointer alignment.

This and the fact that it's a escape hatch for type system (no pointer arithm, no working on data without casting it first, can cast everything to `void *` and from `void *`) makes me a_void_ it all at costs if possible.

IMO for binaries, `uint8_t`/`unsigned char` works well. Anyways, since this is a nitpick / discussion point, I'll leave it as it was.